Significant Read Performance Improvements

rrauch opened the pull request:
I recently noticed `asynchronous-codec` was performing slower than anticipated in a project I'm working on, read operations in particular. To get a better sense of things, I compared its performance against `tokio-codec`, which gave me the throughput I had been expecting for my workload.

After some digging and experimentation, I've ended up with two changes that, when combined, appear to bring `asynchronous-codec`'s read performance effectively on par with `tokio-codec`.

**Here's a quick overview of the changes:**

1.  **Configurable Read Buffer Capacity:** I've introduced a new constructor, `FramedRead::with_capacity`, which allows users to specify the initial size of the internal read buffer. Previously, this was hardcoded to 8KiB. This is similar to what `tokio-codec` does and allows fine tuning for different workloads. 
2.  **Zero-Copy Reads:** The internal read mechanism has been optimized to avoid an unnecessary data copy. Data is now read directly from the underlying `AsyncRead` source into `FramedRead`'s internal `BytesMut` buffer. This eliminates an intermediate allocation and copy for each read operation. However, it does require some `unsafe` Rust, and it relies on the *de-facto* contract that `futures::io::AsyncRead` implementations will only *write* to the provided buffer and not read from its potentially uninitialized parts.

Some benchmarks to illustrate the impact of these changes:

### Performance Benchmarks

The benchmark involves reading a 3 GiB file from fast, local NVMe storage. The `BytesCodec` is used to frame the data.

#### Debug Build Benchmarks

| Version | Buffer Size | Average Throughput (MiB/s) | Diff vs. Unmodified (%) | Diff vs. Tokio-codec (64KiB) (%) |
| :------------ | :---------- | :------------------------- | :-------------------- | :------------------------------- |
| `async-codec` (Unmodified) | 8KiB | 475.76 | 0.0% | -83.8% |
| `async-codec` (configurable buffer) | 8KiB | 500.92 | 5.3% | -82.9% |
| `async-codec` (configurable buffer) | 64KiB | 2297.17 | 382.8% | -21.7% |
| `async-codec` (configurable + zero copy) | 8KiB | 536.70 | 12.8% | -81.7% |
| `async-codec` (configurable + zero copy) | 64KiB | **2932.17** | 516.3% | 0.0% |
| `tokio-codec` | 8KiB | 541.91 | 13.9% | -81.5% |
| `tokio-codec` | 64KiB | 2932.17 | 516.3% | 0.0% |

#### Release Build Benchmarks

| Version | Buffer Size | Average Throughput (MiB/s) | Diff vs. Unmodified (%) | Diff vs. Tokio-codec (64KiB) (%) |
| :------------ | :---------- | :------------------------- | :-------------------- | :------------------------------- |
| `async-codec` (Unmodified) | 8KiB | 841.58 | 0.0% | -78.0% |
| `async-codec` (configurable buffer) | 8KiB | 871.21 | 3.5% | -77.2% |
| `async-codec` (configurable buffer) | 64KiB | 3266.56 | 288.1% | -14.5% |
| `async-codec` (configurable + zero copy) | 8KiB | 886.15 | 5.3% | -76.8% |
| `async-codec` (configurable + zero copy) | 64KiB | **3966.25** | 371.3% | 3.8% |
| `tokio-codec` | 8KiB | 906.72 | 7.7% | -76.3% |
| `tokio-codec` | 64KiB | 3819.52 | 353.8% | 0.0% |

For my workload, reads are now around **500%** faster than before. Overall, `asychronous-codec`'s read performance is now on par with `tokio-codec`.

rrauch committed:
feat(framed-read): Allow configuring initial buffer capacity

Adds `FramedRead::with_capacity` to allow users to specify the initial
internal buffer size. Previously, the buffer was hardcoded to 8KiB,
which could be inefficient for certain workloads. This change enables
users to optimize I/O performance by tuning the buffer capacity.

rrauch committed:
perf(framed-read): Implement zero-copy reads

Optimizes the read path by removing an unnecessary data copy during reads.
Previously, data was read into a temporary buffer before being copied to the final `BytesMut` buffer.
Now, data is written directly into the final, potentially uninitialized buffer, eliminating the intermediate allocation and copy.

mxinden commented:
Thank you for the work @rrauch.

I no longer use `asynchronous-codec` myself, and thus don't actively maintain it.

Maybe you want to create a fork. Happy to link to the various alternatives and archive this project.

//CC @jxs since you are using asynchronous codec as well.

https://github.com/libp2p/rust-libp2p/blob/70082df7e6181722630eabc5de5373733aac9a21/Cargo.lock#L310-L321

jxs commented:
Hi @mxinden thanks for the ping!
Can we then move this repo to the libp2p org and you give publishing rights?

jxs reviewed (commented):
Hi,
thanks for looking into this! Left a comment. 
Can you also share the benchmarks code?
Cheers!

jxs commented on src/framed_read.rs:
```diff
@@ -176,15 +190,40 @@ where
+            //
+            // SAFETY: This code relies on the de-facto contract that `poll_read`
+            // implementations will not read from the buffer before writing to it.
+            let buf = unsafe {
+                let chunk = this.buffer.chunk_mut();
+                std::slice::from_raw_parts_mut(chunk.as_mut_ptr(), chunk.len())
+            };
```
I am a bit uncomfortable with this, [`from_raw_parts_mut`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts_mut.html)'s doc states:
- data must be non-null, [valid](https://doc.rust-lang.org/std/ptr/index.html#safety) for both reads and writes for len * size_of::<T>() many bytes, and it must be properly aligned. 
- data must point to len consecutive properly initialized values of type T.

which corroborates Rust's community idea that referencing uninitialized data is instant UB, see a discussion [here](https://users.rust-lang.org/t/unitialized-memory-safety-question/105073).

Did you benchmark without zero copying? If the performance improvements are negligible I'd rather go with that version, another alternative is zeroing the uninitialized data and using a cursor to the filled data in the buffer, wdyt?

rrauch replied:
I agree with you. I am also not 100% comfortable with it.

It is unfortunate `futures::io::AsyncRead` decided to use a `&mut [u8]` for the read buffer. For high-performance reads this will remain an issue I am afraid. Tokio's decision to use an abstraction allows them to avoid the UB risk here.

However, in situations where the underlying reader is known to be sane it is de-facto safe to use.

I have updated my PR with a compromise that is 100% safe while still using zero-copy semantics. However, it will manually zero-init any new, potentially uninitialized buffer space before passing it to the reader. This comes at a cost of course.

In my (admittedly very unscientific) benchmarks the performance drop is in the 5-10% range. I guess this is just the price to pay for safety in this particular situation.

Overall this is probably the better way to proceed. I have been thinking of adding an *unsafe* option, e.g.

```rust
unsafe fn disable_buffer_initialization(&mut self) ...
```

If set, the zeroing step is simply skipped. For cases where the 5-10% performance drop matter and the readers are known to be sane. Making it an `unsafe` option would mean the user has to be very aware of the trade-offs and risks. 

How would you feel about this?

jxs replied:
thanks for the followup, yeah I think it's reasonable to offer the unsafe version with guard rails, I prefer to feature gate it if you agree, I also think we should explain the conditions in the README.md for being UB free. Meanwhile I'd also like to keep the zeroed version with only safe code, I tinkered with your code and did https://github.com/rrauch/asynchronous-codec/compare/read-performance...jxs:rust-asynchronous-codec:read-performance wdyt?
ping @elenaf9 wdyt as well

rrauch replied:
I've been thinking about this for quite some time now. The issue here is that we must make the decision whether it's safe/sound to use an uninitialized buffer on per-Reader basis. So a general feature flag to enable this is still a risk.

This is a shortcoming of the `AsyncRead` (and also non-async `Read`) trait. I keep seeing this pattern *in the wild* over and over again.

To deal with this, I created the [uninit-read](https://crates.io/crates/uninit-read) crate. Essentially, it's a simple marker trait `UninitRead` that indicates if a given Reader can safely read into an uninitialized buffer.

I've updated my PR to use the `uninit-read` crate to automatically determine if it's safe to use an uninitialized buffer or not.

rrauch committed:
fix: use `spare_capacity_mut` instead of `has_remaining_mut` to determine usable buffer size

rrauch committed:
Additional read check added

rrauch committed:
Initialize buffer spare capacity to prevent UB from uninitialized reads

Zero-initialize newly allocated buffer space before passing to poll_read
to ensure all bytes are safe to access.

rrauch committed:
Updated comments to correctly reflect changes in previous commit

rrauch committed:
Simplified buffer allocation handling and added safety measure

Externally supplied buffers are now safely initialized prior to usage.

rrauch committed:
Added unsafe option to disable buffer initialization

rrauch committed:
Added setter to change buffer capacity & improved docs

rrauch commented:
I made some additional changes to this PR:

* the buffer allocation logic is now simpler and cleaner
* I added buffer initialization to externally supplied buffers. This was an oversight in the prior version.
* added the `set_capacity` method to allow changing the buffer size at any time
* I've added the unsafe `disable_buffer_initialization` option I suggested above

You can now easily compare the performance difference between using uninitialized and initialized buffers.
The outcome will depend highly on the workload, hardware, etc.

rrauch committed:
Updated comment to reflect prior changes

rrauch committed:
Added missing methods to `Framed` and removed superfluous constructor

The `Framed` type was missing the new read-related option setters. With the new setters the `with_capacity` constructor is not really needed or that useful anymore.

rrauch committed:
Inlined framed_read_2_with_capacity

rrauch committed:
`UninitRead` support added to safely allow reading into uninitialized buffers.

rrauch committed:
Removed unsafe `disable_buffer_initialization` option.

Support for uninitialized read buffers is now solely based on whether the inner reader is known to be safe or not.
